### PR TITLE
PE-700 - Add time spent per user report.

### PR DESCRIPTION
**Pull request by Squirrel18**
## Description:

This PR adds the time spent per user report.
Also adds support for additional request data to be included in the report request and overwrite additional request arguments from the command line.

## Configuration:

This report requires the GOOGLE_OAUTH_CREDENTIALS values.

```
"TIME_SPENT_PER_USER_REPORT": {
    "REPORT_URL": "/proversity-reports/api/v0/generate-time-spent-per-user-report",
    "BACKEND_REPORT": "proversity_reports_script.report_backend.time_spent_per_user_report:TimeSpentPerUserReportBackend",
    "EXTRA_REQUEST_DATA": {
        "date": date to request the report YYYY-MM-DD
    },
    "EXTRA_DATA": {
        "BUCKET_NAME": S3 bucket name,
        "SPREADSHEET_DATA": {
            "time_spent_sheet_id_course-id": Google Spreadsheet id
        }
    }
}
```

## Reviewers:

- [ ] @andrey-canon

### CODE REVIEW

File: proversity_reports_script/report_backend/time_spent_per_user_report.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+            user_report_data.update({
+                vertical_name: course_block.get('average_time_spent', 0),
+            })
+
+        report_course_data.append(user_report_data)
+
+    return report_course_data
+
+
+def grouping_report_data_by_key(course_data, grouping_key):
```
Comment (andrey-canon) [inline proversity_reports_script/report_backend/time_spent_per_user_report.py:181]: please improve the performance there are a lot of `for` statements

File: proversity_reports_script/report_backend/time_spent_per_user_report.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+                )
+
+                time_spent_per_cohort_report_data = grouping_report_data_by_key(course_data, 'user_cohort')
+
+                self.create_csv_file(
+                    file_name='{}-cohorts'.format(file_name),
+                    body_dict=time_spent_per_cohort_report_data,
+                    headers=time_spent_per_cohort_report_data[0].keys(),
+                    course_id=course_key,
+                    spreadsheet_id=self.spreadsheet_data.get('time_spent_sheet_id_{}'.format(course_key)),
```
Comment (andrey-canon) [inline proversity_reports_script/report_backend/time_spent_per_user_report.py]: you can get this inside the method

File: proversity_reports_script/report_backend/time_spent_per_user_report.py [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+                    spreadsheet_range_name='Sheet2',
+                    file_name_prefix='teams-',
+                )
+
+                time_spent_per_cohort_report_data = grouping_report_data_by_key(course_data, 'user_cohort')
+
+                self.create_csv_file(
+                    file_name='{}-cohorts'.format(file_name),
+                    body_dict=time_spent_per_cohort_report_data,
+                    headers=time_spent_per_cohort_report_data[0].keys(),
```
Comment (andrey-canon) [inline proversity_reports_script/report_backend/time_spent_per_user_report.py]: you can get this inside the method by using the body_dict

**Review by andrey-canon — CHANGES_REQUESTED** [review]

File: proversity_reports_script/report_backend/time_spent_per_user_report.py (hunk 3 above)
Comment (Squirrel18) [reply proversity_reports_script/report_backend/time_spent_per_user_report.py]: Addressed. @andrey-canon

File: proversity_reports_script/report_backend/time_spent_per_user_report.py (hunk 2 above)
Comment (Squirrel18) [reply proversity_reports_script/report_backend/time_spent_per_user_report.py]: Addressed. @andrey-canon

File: proversity_reports_script/report_backend/time_spent_per_user_report.py (hunk 1 above)
Comment (Squirrel18) [reply proversity_reports_script/report_backend/time_spent_per_user_report.py:181]: Addressed. @andrey-canon

**Review by andrey-canon — APPROVED** [review]

**Pull request merged**